resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -111,66 +109,104 @@ private[spark] class Client(
+                      val driverLauncher = getDriverLauncherClient(kubernetesClient, service)
+                      val ping = Retry.retry(5, 5.seconds) {
+                        driverLauncher.ping()
+                      }
+                      ping onFailure {
+                        case t: Throwable =>
                           submitCompletedFuture.setException(t)
+                          kubernetesClient.services().delete(service)
+                      }
+                      val submitComplete = ping.flatMap { _ =>
+                        Future {
+                          sparkConf.set("spark.driver.host", pod.getStatus.getPodIP)
+                          val submitRequest = buildSubmissionRequest()
+                          driverLauncher.create(submitRequest)
                         }
-                    }
-                    val submitComplete = ping andThen {
-                      case Success(_) =>
-                        driverLauncher.create(submitRequest)
-                        submitCompletedFuture.set(true)
-                    }
-                    submitComplete onFailure {
-                      case t: Throwable =>
-                        if (!submitCompletedFuture.isDone) {
+                      }
+                      submitComplete onFailure {
+                        case t: Throwable =>
                           submitCompletedFuture.setException(t)
+                          kubernetesClient.services().delete(service)
+                      }
+                      val adjustServicePort = submitComplete.flatMap { _ =>
+                        Future {
+                          // After submitting, adjust the service to only expose the Spark UI
+                          val uiServicePort = new ServicePortBuilder()
+                            .withName(UI_PORT_NAME)
+                            .withPort(uiPort)
+                            .withNewTargetPort(uiPort)
+                            .build()
+                          kubernetesClient.services().withName(kubernetesAppId).edit()
+                            .editSpec()
+                              .withType("ClusterIP")
+                              .withPorts(uiServicePort)
+                              .endSpec()
+                            .done
                         }
-                    }
-                  } catch {
-                    case e: Throwable =>
-                      if (!submitCompletedFuture.isDone) {
+                      }
+                      adjustServicePort onSuccess {
+                        case _ =>
+                          submitCompletedFuture.set(true)
+                      }
+                      adjustServicePort onFailure {
+                        case throwable: Throwable =>
+                          submitCompletedFuture.setException(throwable)
+                          kubernetesClient.services().delete(service)
```

Comment:
There is a `Utils.tryAndLogNonFatal` but I've actually had some problems with that method awhile ago, particularly with nested closures. However it was awhile ago and I can't remember what the exact problem was. Might want to look into that a little more.